pyteal/ast/subroutine.py

```diff
@@ -446,53 +620,93 @@ def evaluateSubroutine(subroutine: SubroutineDefinition) -> SubroutineDeclaratio
         Type 1. (by-value) use ScratchVar.store() to pick the actual value into a local scratch space
         Type 2. (by-reference) ALSO use ScratchVar.store() to pick up from the stack
             NOTE: SubroutineCall.__teal__() has placed the _SLOT INDEX_ on the stack so this is stored into the local scratch space
         Type 3. (ABI) abi_value.stored_value.store() to pick from the stack
+        Type 4. (ABI-output-arg) it is not really used here, since it is only generated internal of the subroutine
 
     Usage (B) "loadedArgs" - Passing through to an invoked PyTEAL subroutine AST:
         Type 1. (by-value) use ScratchVar.load() to have an Expr that can be compiled in python by the PyTEAL subroutine
         Type 2. (by-reference) use a DynamicScratchVar as the user will have written the PyTEAL in a way that satisfies
             the ScratchVar API. I.e., the user will write `x.load()` and `x.store(val)` as opposed to just `x`.
         Type 3. (ABI) use abi_value itself after storing stack value into scratch space.
+        Type 4. (ABI-output-arg) generates a new instance of the ABI value,
+            and appends a return expression of stored value of the ABI keyword value.
     """
 
     def var_n_loaded(
         param: str,
-    ) -> Tuple[ScratchVar, Union[ScratchVar, abi.BaseType, Expr]]:
-        loaded: Union[ScratchVar, abi.BaseType, Expr]
-        argVar: ScratchVar
+    ) -> tuple[ScratchVar, ScratchVar | abi.BaseType | Expr]:
+        _loaded_var: ScratchVar | abi.BaseType | Expr
+        _argument_var: ScratchVar
 
         if param in subroutine.by_ref_args:
-            argVar = DynamicScratchVar(TealType.anytype)
-            loaded = argVar
+            _argument_var = DynamicScratchVar(TealType.anytype)
+            _loaded_var = _argument_var
         elif param in subroutine.abi_args:
             internal_abi_var = subroutine.abi_args[param].new_instance()
-            argVar = internal_abi_var.stored_value
-            loaded = internal_abi_var
+            _argument_var = internal_abi_var.stored_value
+            _loaded_var = internal_abi_var
         else:
-            argVar = ScratchVar(TealType.anytype)
-            loaded = argVar.load()
+            _argument_var = ScratchVar(TealType.anytype)
+            _loaded_var = _argument_var.load()
 
-        return argVar, loaded
+        return _argument_var, _loaded_var
 
     args = subroutine.arguments()
-    argumentVars, loadedArgs = zip(*map(var_n_loaded, args)) if args else ([], [])
+    args = [arg for arg in args if arg not in subroutine.output_kwarg]
+
+    argument_vars: list[ScratchVar] = []
+    loaded_args: list[ScratchVar | Expr | abi.BaseType] = []
+    for arg in args:
+        argument_var, loaded_arg = var_n_loaded(arg)
+        argument_vars.append(argument_var)
+        loaded_args.append(loaded_arg)
+
+    abi_output_kwargs: dict[str, abi.BaseType] = {}
+    if len(subroutine.output_kwarg) > 1:
```

Comment:
I recommend putting this validation at the top so you can "fail fast"

Reply:
good point